samples/nrf9160/nrf_cloud_mqtt_multi_service/src/connection.c

```diff
@@ -0,0 +1,648 @@
+			evt->modem_evt == LTE_LC_MODEM_EVT_SEARCH_DONE ?
+				"Search done" :
+			evt->modem_evt == LTE_LC_MODEM_EVT_RESET_LOOP ?
+				"Reset loop detected" :
+			evt->modem_evt == LTE_LC_MODEM_EVT_BATTERY_LOW ?
+				"Low battery" :
+			evt->modem_evt == LTE_LC_MODEM_EVT_OVERHEATED ?
+				"Modem is overheated" :
+				"Unknown");
+		break;
+	default:
+		break;
+	}
+}
+
+/**
+ * @brief Updates the nRF Cloud shadow with information about supported capabilities, current
+ *  firmware running, FOTA support, and so on.
+ */
+static void update_shadow(void)
+{
+	int err;
+	struct nrf_cloud_svc_info_fota fota_info = {
+		.application = fota_capable(),
+		.bootloader = fota_capable() && boot_fota_capable(),
+		.modem = fota_capable()
+	};
+	struct nrf_cloud_svc_info_ui ui_info = {
+		.gps = true,
+		.temperature = true,
+	};
+	struct nrf_cloud_svc_info service_info = {
+		.fota = &fota_info,
+		.ui = &ui_info
+	};
+	struct nrf_cloud_device_status device_status = {
+		//TODO: What does this do, and do we need to update it?
+		.modem = NULL,
+		.svc = &service_info
+	};
+
+	err = nrf_cloud_shadow_device_status_update(&device_status);
+	if (err) {
+		LOG_ERR("Failed to update device shadow, error: %d", err);
+	}
+}
+
+int consume_device_message(void)
+{
+	char *msg;
+	int ret;
+
+	LOG_DBG("Consuming an enqueued device message.");
+
+	/* Wait until a message is available to send */
+	ret = k_msgq_get(&device_message_queue, &msg, K_FOREVER);
+	if (ret) {
+		LOG_ERR("Failed to retrieve item from outgoing message queue, error: %d", ret);
+		return -ret;
+	}
+
+	/* Wait until we are able to send it */
+	LOG_DBG("Waiting for valid connection before transmitting device message");
+	(void)await_connection(K_FOREVER);
+
+	/* Attempt to send it */
+	/* Note, it is possible (and better) to batch-send device messages when more than one is
+	 * queued up. We limit this sample to sending individual messages mainly to keep the sample
+	 * simple and accessible. See the Asset Tracker V2 application for an example of batch
+	 * message sending.
+	 */
+	LOG_DBG("Attempting to transmit enqueued device message.");
+	struct nrf_cloud_tx_data mqtt_msg = {
+		.data.ptr = msg,
+		.data.len = strlen(msg),
+		.qos = MQTT_QOS_0_AT_MOST_ONCE,
+		.topic_type = NRF_CLOUD_TOPIC_MESSAGE,
+	};
+
+	ret = nrf_cloud_send(&mqtt_msg);
+	if (ret) {
+		LOG_ERR("Transmission of enqueued device message failed, nrf_cloud_send "
+			"gave error: %d", ret);
+	}
+
+	LOG_DBG("Finished consuming enqueued device message.");
+
+	/* Clean up */
+	k_free(msg);
+	return ret;
+}
+
+void consume_device_messages(void)
+{
+	while (true) {
+		(void)consume_device_message();
+
+		/* Wait for a bit to avoid overloading the modem with messages all at once */
+		k_sleep(K_SECONDS(CONFIG_MIN_MSG_INTERVAL_SECONDS));
+	}
+}
+
+int send_device_message(const char *const msg)
+{
+	/* Copy message data onto the heap */
+	/* (will be freed by the message consumer when done) */
+	char *msgbuf = k_malloc(strlen(msg) + 1);
```

Comment:
Good question (and keen observation, regarding callers) -- I felt it would be more instructive to a beginner to demonstrate sending a device message directly, and then build a variant of that function to support cJSON messages (since that's what the sample actually needs). I also felt it might be useful to someone building off of the sample if they decide they want to try sending raw-string device messages. Or do you think that is a sufficiently niche use-case that I might as well combine the two functions?

If I don't combine the two functions, then, I think it would be best practice to keep the swap off to heap included, especially since failure to pass a heap-allocated string, or passing a string that mistakenly gets freed by non-consumer thread could lead to a kernel panic.

In fact, even if I implemented the flag you suggest, I would prefer not to use it, since using it might actually make the memory management in use harder to understand to a beginner (it might not be obvious that we are relying on send_device_message to free whatever is passed to it, if we include the flag). That could be solved with some documentation stating it explicitly, but wherever possible I am aiming for simpler, easier to understand approaches.

Reply:
a message is simultaneously existing in three different places: as a cjson obj, as the printed string cjson string, and as the message for the queue.  seems unnecessary to me, but i don't have strong enough feelings about it.

Reply:
You're definitely not wrong! But if you don't feel strongly, I would like to leave as is, since I feel the added readability is worth the heap cost.

Reply:
I've just made a similar design decision when implementing resend-on-fail.

Rather than selectively free the enqueued device message string, I always free it, and make a new copy onto the heap whenever a message needs to be resent. This makes the code simpler by allowing me to reuse the send_device_message function in its entirety.